spacetime/tests/test_linear_scalar.py

```diff
@@ -133,4 +138,84 @@ def _march():
             self.assertEqual(self.svr.get_so0(0).ndarray.tolist(),
                              svr2.get_so0(0).ndarray.tolist())
 
+
+class LinearScalarGridTestTC(unittest.TestCase):
+    """
+    Compare linear solver's solution wtih exact solution, because solutions of computation field is a vector,
+    therefore compared the norm of difference bewteen solver's solution and exact solution.
+
+    By comparing norm of difference bewteen solver's solution and exact solution we can check if the linear solver is work properly
+    or check if the solver's mathematical model is correct.
+    """
+    
+    logger = logging.getLogger()
+    logger.setLevel(logging.DEBUG)
+    stream_handler = logging.StreamHandler(sys.stdout)
+    logger.addHandler(stream_handler)
+    
+    @staticmethod
+    def _build_solver(resolution):
+        grid = libst.Grid(0, 4 * 2 * np.pi, resolution)
+        cfl = 0.99
+        dx = (grid.xmax - grid.xmin) / grid.ncelm
+        dt = dx * cfl
+        svr = libst.LinearScalarSolver(grid=grid, time_increment=dt)
+
+        # Initialize
+        for e in svr.selms(odd_plane=False):
+            if e.xctr < 2 * np.pi or e.xctr > 4 * np.pi:
+                v = 0
+                dv = 0
+            else:
+                v = np.sin(e.xctr)
+                dv = np.cos(e.xctr)
+            e.set_so0(0, v)
+            e.set_so1(0, dv)
+
+        svr.setup_march()
+
+        return svr
+
+    @staticmethod
+    def _exact_solution(svr, iter_num, gfun):
+        v = []
+        for e in svr.selms(odd_plane=False):
+            x = (
+                e.xctr - iter_num * svr.dt
+            ) % svr.grid.xmax  # dealt with boundary treatment
+            if x < 2 * np.pi or x > 4 * np.pi:
+                v.append(0)
+            else:
+                v.append(gfun(x))
+        return np.array(v)
+
+    def setUp(self):
+        self.resolution = 256
+        self.svr = self._build_solver(self.resolution)
+        self.cycle = range(1001)
+
+    def test_grid_test(self):
+        log_format = '{:^20}|{:^30}|{:^20}|'
+        self.logger.debug(('\n'+log_format).format('Grid number', 'Error', 'Ratio'))
+        _norm = lambda vec, ord: sum(np.abs(vec) ** ord) ** (1 / ord)
+
+        grid_num = [320, 640, 1280, 2560, 5120, 10240, 20480, 40960, 81920, 163840, 327680]
+        dx = []
+        err = []
+        it_num = 20
+        for grid in grid_num:
+            svr = self._build_solver(grid)
+            svr.march_alpha2(it_num)
+            exact_sol = self._exact_solution(svr, it_num, np.sin)
+            dx.append(svr.grid.ncelm)
+            err.append(_norm(
+                svr.get_so0(0).ndarray - exact_sol, 1))
+        
+        self.logger.debug(log_format.format(grid_num[0], err[-1], 0))
+        for i in range(1, len(err)):
```

Comment:
In https://github.com/yungyuc/turgon/pull/32/files/397ac98bf2451203763bb384a21f3813e573f223#r541016811 I asked this to be changed to be tabulated so that in the test case, it doesn't use a loop to test.  By tabulating, I meant to have explicit numbers for the golden.  Here still uses the loop and doesn't look like what I requested.

Why not just use something like the following?

```python
self.assertAlmostEqual(some_number_for_golden1, math.log(abs(err[0]/err[1]) / math.log(abs[dx[0]/dx[1])))
self.assertAlmostEqual(some_number_for_golden2, math.log(abs(err[1]/err[2]) / math.log(abs[dx[1]/dx[2])))
...
```

Thus, when anything is wrong in the convergence results, we know exactly where it errs.